### PR TITLE
feat: Telemetry logger for Fluid Devtools using public packages

### PR DESCRIPTION
**Pull request by alexvy86**
## Description

Supersedes https://github.com/microsoft/FluidFramework/pull/16351, as an implementation of a logger to send usage telemetry for Fluid Devtools that only requires public npm packages.

As-is, this PR does not send telemetry; a valid instrumentationKey is necessary. We're still exploring options on how to inject that value during the build so it's not in source control.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

[AB#5228](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/5228)

### CODE REVIEW

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		const requestInit: RequestInit = {
+			body: telemetryRequestData,
+			method: "POST",
+			headers: payload.headers,
+			credentials: "include",
+		};
+		fetch(payload.urlString, requestInit)
+			.then((response) => {
+				const headerMap: { [key: string]: string } = {};
+				// response.headers is not a run-of-the-mill array, it satisfies a particular interface that
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:27]: Thanks for the docs here! This is really valuable!

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+			},
+		};
+
+		if ((coreConfig.instrumentationKey ?? "") !== "") {
+			this.enabled = true;
+			this.appInsightsCore.initialize(coreConfig, []);
+		}
+	}
+
+	public send(event: ITelemetryBaseEvent): void {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:91]:
```suggestion
	/**
	 * {@inheritDoc @fluid-experimental/devtools-view#ITelemetryBaseLogger.send}
	 */
	public send(event: ITelemetryBaseEvent): void {
```

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	},
+};
+/**
+ * Logger that sends logs to the OneDS collection endpoint.
+ */
+export class OneDSLogger implements ITelemetryBaseLogger {
+	private readonly appInsightsCore = new AppInsightsCore();
+	private readonly postChannel: PostChannel = new PostChannel();
+
+	private readonly enabled: boolean = false;
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:63]: Can we add some docs to this to explain how it's set / what it means for the end-user?

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	public send(event: ITelemetryBaseEvent): void {
+		if (!this.enabled) {
+			return;
+		}
+
+		// Note: some APIs might fail if the last part of the eventName is not uppercase
+		const category = event.category
+			? `${event.category.charAt(0).toUpperCase()}${event.category.slice(1)}`
+			: "Generic";
+		const eventType = `Office.Fluid.Devtools.${category}`;
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:100]: Nit: Might be worth documenting what this naming pattern accomplishes. And I'm assuming the "Office." prefix is a requirement?

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+			this.appInsightsCore.initialize(coreConfig, []);
+		}
+	}
+
+	public send(event: ITelemetryBaseEvent): void {
+		if (!this.enabled) {
+			return;
+		}
+
+		// Note: some APIs might fail if the last part of the eventName is not uppercase
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:96]: Our APIs? Or 1ds's?

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+	public constructor() {
+		const channelConfig: IChannelConfiguration = {
+			alwaysUseXhrOverride: true,
+			httpXHROverride: fetchHttpXHROverride,
+		};
+
+		// Configure App insights core to send to collector
+		const coreConfig: IExtendedConfiguration = {
+			instrumentationKey: "<KEY-GOES-HERE>",
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:73]: I'm assuming we plan to deal with this in a future PR? Should we leave a TODO comment explaining the plan?

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+			endpointUrl: "https://mobile.events.data.microsoft.com/OneCollector/1.0",
+			loggingLevelConsole: 0, // Do not log to console
+			disableDbgExt: true, // Small perf optimization
+			extensions: [
+				this.postChannel, // Removing this as part of telemetry opt-in would be ideal
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:78]: I'm not sure I follow the comment. Could we elaborate a bit?

File: packages/tools/devtools/devtools-browser-extension/package.json [hunk 8]
```diff
@@ -68,9 +68,14 @@
...
 		"@fluid-experimental/devtools-view": "workspace:~",
+		"@fluidframework/common-definitions": "^0.20.1",
 		"@fluidframework/common-utils": "^1.1.1",
+		"@fluidframework/core-interfaces": "workspace:~",
+		"@microsoft/1ds-core-js": "^3.2.13",
+		"@microsoft/1ds-post-js": "^3.2.13",
 		"react": "^17.0.1",
-		"react-dom": "^17.0.1"
+		"react-dom": "^17.0.1",
+		"tslib": "^1.10.0"
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-browser-extension/package.json:78]: Is this a peer of one of the 1ds libraries or something?

**Review by Josmithr — APPROVED** [review]
This looks great! I love that we'll get to use this without needing MS-internal libraries!

I left a handful of questions and nitpicks, but otherwise looks great to me!

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts (hunk 4 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:100]: I'm still not 100% clear on this myself; I think the whole of "Office.Fluid" is required. I can add some docs to at least ensure that this isn't changed without proper consideration.

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts (hunk 5 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:96]: 1DS (or things that it abstract us from), I'll clarify. Took this from existing logger implementations in partner repos, fwiw :)

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts (hunk 6 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:73]: Yep, I'll add it. It's the task that @chentong7 is taking on now to inject it at build time. Also, thanks for making me catch that I left the placeholder here and not an empty string, which is what the `enabled` private flag checks for. And we should keep this in mind for Tong's task; if we need a placeholder for a build-time replacement, we need it to interact well with the "no key was provided" idea for `enabled`.

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts (hunk 7 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:78]: If we reinstantiated the logger object in response to a change in the telemetry opt-in setting (from the Settings view), we could not provide this channel/sink at all, further ensuring that no telemetry could be sent. It was a bit more cumbersome than the private `enabled` flag, but we could refactor into that later. I didn't think it was absolutely necessary from the start.

File: packages/tools/devtools/devtools-browser-extension/package.json (hunk 8 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/package.json:78]: Yep.

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+			alwaysUseXhrOverride: true,
+			httpXHROverride: fetchHttpXHROverride,
+		};
+
+		// Configure App insights core to send to collector
+		const coreConfig: IExtendedConfiguration = {
+			// A non-empty instrumentation key needs to provided for the logger to do anything.
+			// AB#5167 tracks how to inject one at build time to keep it out of source control.
+			instrumentationKey: "",
+			endpointUrl: "https://mobile.events.data.microsoft.com/OneCollector/1.0",
```
Comment (chentong7) [inline packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:81]: Where do we get this endpoint url? Why is it "mobile"? :raised_eyebrow:

File: packages/tools/devtools/devtools-view/api-report/devtools-view.api.md [hunk 10]
```diff
@@ -18,6 +18,7 @@ export function DevtoolsPanel(props: DevtoolsPanelProps): React_2.ReactElement;
 // @public
 export interface DevtoolsPanelProps {
     messageRelay: IMessageRelay;
+    unloadCallback?: () => void;
```
Comment (chentong7) [inline packages/tools/devtools/devtools-view/api-report/devtools-view.api.md:21]: Do we need to include a logger?.flush() also here to clean up logger if devtool close?

**Review by chentong7 — APPROVED** [review]

File: packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts (hunk 9 above)
Comment (alexvy86) [reply packages/tools/devtools/devtools-browser-extension/src/devtools/TelemetryLogging.ts:81]: `browser.events.data.microsoft.com` works and is even the default so I'm just removing this from here :)

**Pull request merged**